src/js_tests/wirecloud/ConnectionSpec.js

```diff
@@ -0,0 +1,251 @@
+                connection.establish();
+                connection.detach();
+
+                expect(connection.established).toBe(false);
+                expect(callback.calls.count()).toEqual(1);
+            });
+
+            it("should ignore if connection is not established", function () {
+                var connection = new ns.Connection(wiringEngine, sourceEndpoint, targetEndpoint);
+
+                var callback = jasmine.createSpy("callback");
+
+                connection.addEventListener("detach", callback);
+                connection.detach();
+
+                expect(callback.calls.count()).toEqual(0);
+            });
+        });
+
+        describe("remove()", function () {
+
+            it("should remove after disconnecting endpoints", function () {
+                var connection = new ns.Connection(wiringEngine, sourceEndpoint, targetEndpoint);
+
+                var callback1 = jasmine.createSpy("callback1");
+                var callback2 = jasmine.createSpy("callback2");
+
+                connection.addEventListener("detach", callback1);
+                connection.addEventListener("remove", callback2);
+
+                connection.establish();
+                connection.remove();
+
+                expect(connection.established).toBe(false);
+                expect(callback1.calls.count()).toEqual(1);
+                expect(callback2.calls.count()).toEqual(1);
+            });
+        });
+
+        describe("updateEndpoint(endpoint)", function () {
+
+            it("should change the missing source-endpoint", function () {
+                var missingSourceEndpoint = new ns.GhostSourceEndpoint(widgetModel, "missing");
+                var connection = new ns.Connection(wiringEngine, missingSourceEndpoint, targetEndpoint);
+
+                var callback = jasmine.createSpy("callback");
+
+                connection.addEventListener("establish", callback);
+                connection.updateEndpoint(sourceEndpoint);
+
+                expect(connection.established).toBe(true);
+                expect(connection.source).toEqual(sourceEndpoint);
+                expect(callback.calls.count()).toEqual(1);
+            });
+
+            it("should change the missing target-endpoint", function () {
+                var missingTargetEndpoint = new ns.GhostTargetEndpoint(operatorModel, "missing");
+                var connection = new ns.Connection(wiringEngine, sourceEndpoint, missingTargetEndpoint);
+
+                var callback = jasmine.createSpy("callback");
+
+                connection.addEventListener("establish", callback);
+                connection.updateEndpoint(targetEndpoint);
+
+                expect(connection.established).toBe(true);
+                expect(connection.target).toEqual(targetEndpoint);
+                expect(callback.calls.count()).toEqual(1);
+            });
+
+            it("should throw error if endpoint is not valid", function () {
```

Comment:
Would be great to test more invalid values. Please create a describe section with a "throws a TypeError exception if endpoint is not valid" and create a test using `null`,  a number, a string and an object (as minimum).